File: apps/dashboard/src/api/brainstorm.ts
import { api } from "./client";

interface ResponseEnvelope<T> {
  status: "success";
  data: T;
  meta: { request_id: string; timestamp: string };
}

export interface BrainstormQuestion {
  id: string;
  category: string;
  prompt: string;
  required: boolean;
  priority: string;
  answer: string | null;
  status: string;
}

export interface BrainstormMessage {
  id: string;
  role: "assistant" | "user";
  content: string;
  created_at: string;
}

export interface BrainstormSummary {
  readiness_score: number;
  ready_for_pipeline: boolean;
  blockers: string[];
  recommended_preset: string;
  recommended_next_step: string;
  open_questions: number;
  answered_questions: number;
  required_questions_remaining: number;
}

export interface BrainstormSession {
  session_id: string;
  project_id: string;
  status: string;
  started_at: string;
  updated_at: string;
  overview: string;
  refined_brief: string;
  questions: BrainstormQuestion[];
  messages: BrainstormMessage[];
  summary: BrainstormSummary;
  source_context: Record<string, unknown> | null;
  agent_output: Record<string, unknown> | null;
}

export const brainstormApi = {
  start(projectId: string) {
    return api.post<ResponseEnvelope<BrainstormSession>>(
      `/api/v1/projects/${projectId}/brainstorm/start`,
    );
  },

  respond(projectId: string, body: { content: string; question_id?: string }) {
    return api.post<ResponseEnvelope<BrainstormSession>>(
      `/api/v1/projects/${projectId}/brainstorm/respond`,
      body,
    );
  },

  finalize(projectId: string) {
    return api.post<ResponseEnvelope<BrainstormSession>>(
      `/api/v1/projects/${projectId}/brainstorm/finalize`,
    );
  },
};

File: apps/server/src/codebot/api/schemas/brainstorm.py
"""Schemas for project brainstorm session endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class BrainstormQuestion(BaseModel):
    """A guided clarification question for project intake."""

    id: str
    category: str
    prompt: str
    required: bool = True
    priority: str = "high"
    answer: str | None = None
    status: str = "open"


class BrainstormMessage(BaseModel):
    """A message in the brainstorm conversation."""

    id: str
    role: str
    content: str
    created_at: datetime


class BrainstormSummary(BaseModel):
    """Machine-usable summary of brainstorm readiness."""

    readiness_score: int = Field(ge=0, le=100)
    ready_for_pipeline: bool
    blockers: list[str]
    recommended_preset: str
    recommended_next_step: str
    open_questions: int = Field(ge=0)
    answered_questions: int = Field(ge=0)
    required_questions_remaining: int = Field(ge=0)


class BrainstormSessionResponse(BaseModel):
    """Current brainstorm session state for a project."""

    session_id: str
    project_id: UUID
    status: str
    started_at: datetime
    updated_at: datetime
    overview: str
    refined_brief: str
    questions: list[BrainstormQuestion]
    messages: list[BrainstormMessage]
    summary: BrainstormSummary
    source_context: dict[str, Any] | None = None
    agent_output: dict[str, Any] | None = None


class BrainstormRespondRequest(BaseModel):
    """Request body for answering a brainstorm question."""

    content: str = Field(min_length=1)
    question_id: str | None = None